perception/lidar_centerpoint_tvm/lib/preprocess/generate_features.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright 2021-2022 AutoCore Ltd., TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "lidar_centerpoint_tvm/preprocess/generate_features.hpp"
+
+#include <lidar_centerpoint_tvm/utils.hpp>
+
+#include <thread>
+
+namespace
+{
+const std::size_t THREAD_NUM_VFE = 4;
+}  // namespace
+
+namespace autoware
+{
+namespace perception
+{
+namespace lidar_centerpoint_tvm
+{
+
+void generateFeatures_worker(
+  const std::vector<float32_t> & voxel_features, const std::vector<float32_t> & voxel_num_points,
+  const std::vector<int32_t> & coords, const std::size_t num_voxels,
+  const CenterPointConfig & config, std::vector<float32_t> & features, std::size_t thread_idx,
+  std::size_t pillars_per_thread)
+{
+  for (std::size_t idx = 0; idx < pillars_per_thread; idx++) {
+    std::size_t pillar_idx = thread_idx * pillars_per_thread + idx;
+    if (pillar_idx >= num_voxels) return;
+
+    // voxel/pillar information
+    float32_t points_sum[3] = {0.0, 0.0, 0.0};  // sum of x, y, z in the voxel
+    int32_t coordinate[3] = {
+      coords[pillar_idx], coords[pillar_idx + 1],
+      coords[pillar_idx + 2]};                                // 3D position(z,y,x) of the voxel
```

Comment:
It should be `coords[pillar_idx * 3], coords[pillar_idx * 3 + 1], coords[pillar_idx * 3 + 2]`

Reply:
you are right, thanks!